File: c/02_流程控制/02_循环语句.c
/* ============================================================
 * C 语言流程控制 - 循环语句
 * 编译运行：gcc 02_循环语句.c -o 02_循环语句 && ./02_循环语句
 * ============================================================ */

#include <stdio.h>

int main() {
    printf("=== 1. for 循环 ===\n");
    for (int i = 0; i < 5; i++) {
        printf("第 %d 次循环\n", i + 1);
    }

    printf("\n=== 2. for 循环 - 步长 ==\n");
    printf("1到9的奇数: ");
    for (int i = 1; i < 10; i += 2) {
        printf("%d ", i);
    }
    printf("\n");

    printf("\n=== 3. while 循环 ===\n");
    int count = 1;
    while (count <= 5) {
        printf("第 %d 次循环\n", count);
        count++;
    }

    printf("\n=== 4. do-while 循环 ===\n");
    int num = 1;
    do {
        printf("num = %d\n", num);
        num++;
    } while (num <= 5);

    printf("\n=== 5. break 语句 ===\n");
    for (int i = 0; i < 10; i++) {
        if (i == 5) {
            break;
        }
        printf("%d ", i);
    }
    printf("\n遇到 break，循环结束\n");

    printf("\n=== 6. continue 语句 ===\n");
    for (int i = 0; i < 10; i++) {
        if (i % 2 == 0) {
            continue;
        }
        printf("%d ", i);
    }
    printf("\n只打印了奇数\n");

    printf("\n=== 7. goto 语句 ===\n");
    int x = 0;
start:
    printf("x = %d\n", x);
    x++;
    if (x < 3) {
        goto start;
    }

    printf("\n=== 8. 嵌套循环 ===\n");
    printf("九九乘法表:\n");
    for (int i = 1; i < 10; i++) {
        for (int j = 1; j <= i; j++) {
            printf("%dx%d=%d\t", j, i, i * j);
        }
        printf("\n");
    }

    printf("\n=== 9. 使用示例：累加求和 ===\n");
    int total = 0;
    for (int i = 1; i <= 100; i++) {
        total += i;
    }
    printf("1到100的和: %d\n", total);

    printf("\n=== 10. 使用示例：查找质数 ===\n");
    int is_prime(int n) {
        if (n < 2) return 0;
        for (int i = 2; i * i <= n; i++) {
            if (n % i == 0) return 0;
        }
        return 1;
    }

    printf("1到50之间的质数:\n");
    for (int num2 = 1; num2 <= 50; num2++) {
        if (is_prime(num2)) {
            printf("%d ", num2);
        }
    }
    printf("\n");

    printf("\n=== 11. 使用示例：斐波那契数列 ===\n");
    printf("斐波那契数列前10项:\n");
    int a = 0, b = 1;
    for (int i = 0; i < 10; i++) {
        printf("%d ", a);
        int next = a + b;
        a = b;
        b = next;
    }
    printf("\n");

    printf("\n=== 12. for 循环的灵活用法 ===\n");
    // 初始化部分可以有多个变量
    int i, j;
    for (i = 0, j = 10; i < j; i++, j--) {
        printf("i=%d, j=%d\n", i, j);
    }

    // 无限循环
    printf("\n无限循环（用 break 退出）:\n");
    int k = 0;
    for (;;) {
        if (k >= 3) break;
        printf("k = %d\n", k);
        k++;
    }

    return 0;
}

File: go/02_流程控制/02_循环语句.go
// ============================================================
// Go 流程控制 - 循环语句
// 运行方式：go run 02_循环语句.go
// ============================================================

package main

import "fmt"

func main() {
	fmt.Println("=== 1. 基本 for 循环 ===")
	for i := 0; i < 5; i++ {
		fmt.Printf("第 %d 次循环\n", i+1)
	}

	fmt.Println("\n=== 2. 类似 while 的 for 循环 ===")
	count := 1
	for count <= 5 {
		fmt.Printf("count = %d\n", count)
		count++
	}

	fmt.Println("\n=== 3. 无限循环（用 break 退出） ===")
	n := 0
	for {
		if n >= 3 {
			break
		}
		fmt.Printf("无限循环 n = %d\n", n)
		n++
	}

	fmt.Println("\n=== 4. for-range 遍历 ===")

	// 遍历数组/切片
	fmt.Println("--- 遍历切片 ---")
	nums := []int{10, 20, 30, 40, 50}
	for i, v := range nums {
		fmt.Printf("索引 %d: %d\n", i, v)
	}

	// 只取值
	fmt.Println("\n--- 只取值（忽略索引） ---")
	for _, v := range nums {
		fmt.Printf("值: %d\n", v)
	}

	// 遍历字符串
	fmt.Println("\n--- 遍历字符串 ---")
	str := "Hello 世界"
	for i, ch := range str {
		fmt.Printf("索引 %d: %c\n", i, ch)
	}

	// 遍历 map
	fmt.Println("\n--- 遍历 map ---")
	m := map[string]int{"a": 1, "b": 2, "c": 3}
	for k, v := range m {
		fmt.Printf("%s: %d\n", k, v)
	}

	fmt.Println("\n=== 5. break 和 continue ===")

	fmt.Print("break 示例（到5停止）: ")
	for i := 0; i < 10; i++ {
		if i == 5 {
			break
		}
		fmt.Printf("%d ", i)
	}
	fmt.Println()

	fmt.Print("continue 示例（跳过偶数）: ")
	for i := 0; i < 10; i++ {
		if i%2 == 0 {
			continue
		}
		fmt.Printf("%d ", i)
	}
	fmt.Println()

	fmt.Println("\n=== 6. 嵌套循环 ===")
	fmt.Println("乘法表:")
	for i := 1; i <= 5; i++ {
		for j := 1; j <= i; j++ {
			fmt.Printf("%d*%d=%-2d ", j, i, i*j)
		}
		fmt.Println()
	}

	fmt.Println("\n=== 7. 使用示例：查找质数 ===")
	fmt.Println("1 到 50 之间的质数:")
	for num := 1; num <= 50; num++ {
		if isPrime(num) {
			fmt.Printf("%d ", num)
		}
	}
	fmt.Println()

	fmt.Println("\n=== 8. 使用示例：斐波那契数列 ===")
	fmt.Println("前 10 项:")
	for i := 0; i < 10; i++ {
		fmt.Printf("%d ", fibonacci(i))
	}
	fmt.Println()
}

func isPrime(n int) bool {
	if n < 2 {
		return false
	}
	for i := 2; i*i <= n; i++ {
		if n%i == 0 {
			return false
		}
	}
	return true
}

func fibonacci(n int) int {
	if n <= 1 {
		return n
	}
	a, b := 0, 1
	for i := 2; i <= n; i++ {
		a, b = b, a+b
	}
	return b
}
